apps/designer/app/designer/features/style-panel/controls/text/text-control.tsx

```diff
@@ -15,37 +39,195 @@ export const TextControl = ({
-      onChangeComplete={(value) => {
-        console.log("change complete", value);
-        handleChange(value, false);
+      renderTextField={({ inputProps, toggleProps }) => {
+        return (
+          <Box css={{ position: "relative" }}>
+            <TextField
+              {...inputProps}
+              inputRef={inputRef}
+              {...(value.type === "unit" && {
+                onPointerDown: (event: PointerEvent<HTMLElement>) => {
+                  event.preventDefault();
+                },
+                onPointerUp: () => {
+                  if (Object(inputRef.current)["preventFocus"])
+                    Object(inputRef.current)["preventFocus"] = undefined;
+                  else inputRef.current?.focus();
+                },
+                onPointerEnter: (event: PointerEvent<HTMLElement>) => {
+                  if (
+                    inputRef.current?.ownerDocument.activeElement ===
+                    inputRef.current
+                  )
+                    return;
+                  Object(inputRef.current)["scrubObject"] ??=
+                    numericScrubControl(event.currentTarget, {
+                      initialValue: value.value,
+                      onValueInput: (event) => {
+                        Object(inputRef.current).value = String(event.value);
+                        Object(inputRef.current)["preventFocus"] = true;
+                        handleChange("value", event.value, true);
+                      },
+                      onValueChange: (event) => {
+                        setValue(String(event.value));
+                      },
+                    });
+                },
+              })}
+              onFocus={(event: FocusEvent<HTMLInputElement>) => {
+                event.currentTarget.select();
+                Object(inputRef.current)["scrubObject"]?.disconnectedCallback();
+                Object(inputRef.current)["scrubObject"] = undefined;
+              }}
+              onBlur={(event) => {
+                if (
+                  inputRef.current?.value !==
+                  inputRef.current?.getAttribute("value")
+                )
+                  setValue(event.currentTarget.value);
+              }}
+              onKeyDown={(event: KeyboardEvent<HTMLInputElement>) => {
+                if (handleKeyDown(event)) inputProps?.onKeyDown?.(event);
+              }}
+              prefix={
+                <PropertyIcon
+                  property={styleConfig.property}
+                  label={styleConfig.label}
+                />
+              }
+              suffix={
+                <IconButton
+                  {...(value.type !== "unit" && toggleProps)}
+                  css={{
+                    visibility: items.length ? "visible" : "hidden",
+                    position: "absolute",
+                    right: "1px",
+                    top: "1px",
+                    width: "auto",
+                    height: "calc(100% - 2px)",
```

Comment:
again, we said no nudges

Reply:
positioning should be done via layout
sizes should be done using token values, if there is none that fits the exact pixel value we need - take the next closest, we have to fix tokens, not workaround with nudges